### PR TITLE
Added Composite Pattern

### LINKED ISSUE #18: Composite design pattern
We need an implementation for this pattern. It would be best if it is ready before Friday 31 so we can teach in class 😏

### PR DESCRIPTION
**Pull request by Cotel**
This close #18 

Don't mind the commit name, I forgot to change it 😝 

The composite pattern is useful to represent a group of objects as one. Don't know if this example is enough but I couln't think of a more complex example where this can be useful.

It would be great if somebody told if there are examples of everyday code where we are using this pattern (for example, every DAO is a Facade. I would like to see something like that for Composite)

### CODE REVIEW

File: src/main/kotlin/Composite/base/Kitchen.kt [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+
+class Kitchen(private var cookers: MutableList<Cooker> = mutableListOf()) : Cooker {
+  override fun cook() {
+    cookers.forEach {it.cook()}
+  }
+
+  fun add(cooker: Cooker) = cookers.add(cooker)
+
+  fun remove(cooker: Cooker) {
+    cookers = cookers.filterNot { it == cooker } as MutableList<Cooker>
```
Comment (Martrout96) [inline src/main/kotlin/Composite/base/Kitchen.kt]: why you don't use `cookers.remove(cooker)` ??

Comment (Cotel) [inline src/main/kotlin/Composite/base/Kitchen.kt]: Omg, I was complicating myself too much 😨 

You can change it yourself and learn how to collaborate on an PR 😉

Comment (tonilopezmr) [inline src/main/kotlin/Composite/base/Kitchen.kt]: LOL @Martrout96

File: src/main/kotlin/Composite/base/Kitchen.kt [hunk 2]
```diff
@@ -8,6 +8,6 @@ class Kitchen(private var cookers: MutableList<Cooker> = mutableListOf()) : Cook
   fun add(cooker: Cooker) = cookers.add(cooker)
 
   fun remove(cooker: Cooker) {
-    cookers = cookers.filterNot { it == cooker } as MutableList<Cooker>
+    cookers.remove(cooker)
```
Comment (Cotel) [inline src/main/kotlin/Composite/base/Kitchen.kt]: How further can I go down...

Comment (Martrout96) [inline src/main/kotlin/Composite/base/Kitchen.kt]: you are a silly person

File: src/main/kotlin/Composite/Main.kt [hunk 3]
```diff
@@ -0,0 +1,19 @@
+package Composite
+
+import Composite.base.ChineseCooker
+import Composite.base.ItalianCooker
+import Composite.base.Kitchen
+
+fun main(args : Array<String>) {
```
Comment (tonilopezmr) [inline src/main/kotlin/Composite/Main.kt]: Stop using Main, We can see how it used in Test! 

That you are trying to test in your main, test it in unit tests!!

File: src/main/kotlin/Composite/base/ChineseCooker.kt [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package Composite.base
```
Comment (tonilopezmr) [inline src/main/kotlin/Composite/base/ChineseCooker.kt]: Stop using `PatternName.base`  it's very annoying please.

**Comment by Cotel** [comment]
I was confused with this pattern because I found 2 examples that were doing different things.

[This is one](https://es.wikipedia.org/wiki/Composite_(patr%C3%B3n_de_dise%C3%B1o). In here, they use the pattern to treat a group of objects as one and operate with them as a group.

[This is the other one](https://www.tutorialspoint.com/design_pattern/composite_pattern.htm). In here, they use the pattern to establish a hierarchy in one object. This time, it is only for representing data, it hasn't any behaviour. I think this example doesn't need a pattern, this could be easily done with a Tree data structure.

It is confusing because they are alike but they do not do the same things. What is the real purpose of this pattern?

**Comment by tonilopezmr** [comment]
I don't like first example because are broken liskov principle. In `Hoja` class. The [english versión](https://en.wikipedia.org/wiki/Composite_pattern)  has better implementation, because `Leaf` doesn't not have `add` function 👍

File: src/test/kotlin/Composite/CompositeTest.kt [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    val kitchen = MockKitchen()
+    assertThat(kitchen.cookers.size, `is`(0))
+
+    val italian = ItalianCooker()
+    kitchen.add(italian)
+    assertThat(kitchen.cookers.size, `is`(1))
+  }
+
+  @Test
+  fun `composite should allow to remove childrens`() {
```
Comment (tonilopezmr) [inline src/test/kotlin/Composite/CompositeTest.kt]: Seriously you are testing that you are implemented a good `MockKitchen` when calls kitchen.remove(italian) it calls list.remove

File: src/test/kotlin/Composite/CompositeTest.kt [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+  @Test
+  fun `times cooked should be equal to cookers length after cooking with all childrens of composite`() {
+    val kitchen = MockKitchen()
+    val italian = ItalianCooker()
+    val chinese = ChineseCooker()
+    kitchen.add(italian)
+    kitchen.add(chinese)
+
+    kitchen.cook()
+    assertThat(kitchen.timesCooked, `is`(kitchen.cookers.size))
```
Comment (tonilopezmr) [inline src/test/kotlin/Composite/CompositeTest.kt]: this seems better but.. It's the same as before, you are testing if you are call to cook in your `MockKitchen`

**Comment by Cotel** [comment]
To be honest I didn't know how to test this at all because I think I am not using the pattern correctly. Even the Composite in Chain of Responsability is better 🤔

**Comment by tonilopezmr** [comment]
I will think

**Comment by tonilopezmr** [comment]
@Cotel You can see how it's does that you wanted to do 👍

**Comment by tonilopezmr** [comment]
@Cotel You can see how it's does that you wanted to do 👍

**Comment by tonilopezmr** [comment]
@Cotel You can see how it's does that you wanted to do 👍

**Comment by Cotel** [comment]
How it's done*.

Reviewing now 👍

File: src/test/kotlin/oop/Composite/CompositeTest.kt [hunk 7]
```diff
@@ -1,59 +1,64 @@
...
+    val cookers = mutableListOf<Cooker>()
+    val kitchen = Kitchen(cookers)
+    val italian = MockCooker()
+    val chinese = MockCooker()
     kitchen.add(italian)
     kitchen.add(chinese)
 
     kitchen.cook()
-    assertThat(kitchen.timesCooked, `is`(kitchen.cookers.size))
+    assertThat(italian.times + italian.times, `is`(cookers.size))
```
Comment (Cotel) [inline src/test/kotlin/oop/Composite/CompositeTest.kt]: Why italian twice? You meant `italian.times + chinese.times`?

File: src/test/kotlin/oop/Composite/CompositeTest.kt [hunk 8]
```diff
@@ -43,12 +43,14 @@ class CompositeTest {
...
     val italian = MockCooker()
+    val another = MockCooker()
     val chinese = MockCooker()
     kitchen.add(italian)
+    kitchen.add(another)
     kitchen.add(chinese)
 
     kitchen.cook()
-    assertThat(italian.times + italian.times, `is`(cookers.size))
+    assertThat(italian.times + chinese.times, `is`(cookers.size))
```
Comment (Cotel) [inline src/test/kotlin/oop/Composite/CompositeTest.kt]: Nice 😉

File: src/main/kotlin/oop/Composite/Cooker.kt [hunk 9]
```diff
@@ -0,0 +1,57 @@
+package oop.Composite
+
+interface Cooker {
+  fun cook()
+}
+
+class ChineseCooker(private val cookers: MutableList<Cooker> = mutableListOf()) : Cooker {
```
Comment (Cotel) [inline src/main/kotlin/oop/Composite/Cooker.kt]: Why this one and ItalianCooker have a list of cookers now? 🤔

File: src/test/kotlin/oop/Composite/CompositeTest.kt (hunk 7 above)
Comment (tonilopezmr) [inline src/test/kotlin/oop/Composite/CompositeTest.kt]: is done here 2d9b6ee1acad83a03e5653273cc538281588638d

File: src/main/kotlin/oop/Composite/Cooker.kt (hunk 9 above)
Comment (tonilopezmr) [inline src/main/kotlin/oop/Composite/Cooker.kt]: Because Is a branch that could has more `Leafs`

Comment (tonilopezmr) [inline src/main/kotlin/oop/Composite/Cooker.kt]: You can see an example here https://github.com/CloudCoders/Design-Patterns/pull/23/commits/016e9345fcd9bd4e8be8d53f87befb97dcf5e0e9#diff-2d06d22093fdc986cc4767c7cb391508R13

**Pull request merged**